oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java

```diff
@@ -0,0 +1,813 @@
+            return super.createOrUpdate(collection, update);
+        }
+        
+        private <T extends Document> List<T> superCreateOrUpdate(Collection<T> collection, List<UpdateOp> updateOps) {
+            List<T> result = new ArrayList<T>(updateOps.size());
+            for (UpdateOp update : updateOps) {
+                result.add(super.createOrUpdate(collection, update));
+            }
+            return result;
+        }
+        
+        @Override
+        public <T extends Document> List<T> createOrUpdate(Collection<T> collection,
+                List<UpdateOp> updateOps) {
+            final List<UpdateOp> trimmedOps = new LinkedList<>();
+            final List<UpdateOp> lateOps = new LinkedList<>();
+            trimOps(collection, updateOps, trimmedOps, lateOps);
+            List<T> result = superCreateOrUpdate(collection, trimmedOps);
+            registerWrittenUpdateOp(collection, trimmedOps);
+            if (trimmedOps.size() != updateOps.size()) {
+                breakpoint("wanted " + updateOps.size() + " ops, allowed only " + trimmedOps.size());
+                superCreateOrUpdate(collection, lateOps);
+            }
+            return result;
+        }
+        
+        @Override
+        public <T extends Document> T findAndUpdate(Collection<T> collection,
+                UpdateOp update) {
+            allowOp(collection);
+            registerWrittenUpdateOp(collection, update);
+            return super.findAndUpdate(collection, update);
+        }
+        
+        @Override
+        public <T extends Document> void remove(Collection<T> collection,
+                List<String> keys) {
+            throw new IllegalStateException("not yet implemented");
+//            super.remove(collection, trimString(collection, keys));
+        }
+        
+        @Override
+        public <T extends Document> int remove(Collection<T> collection,
+                Map<String, Long> toRemove) {
+            throw new IllegalStateException("not yet implemented");
+        }
+        
+        @Override
+        public <T extends Document> void remove(Collection<T> collection, String key) {
+            allowOp(collection);
+            super.remove(collection, key);
+        }
+        
+        @Override
+        public <T extends Document> int remove(Collection<T> collection,
+                String indexedProperty, long startValue, long endValue)
+                throws DocumentStoreException {
+            throw new IllegalStateException("not yet implemented");
+        }
+    }
+
+    @Rule
+    public DocumentMKBuilderProvider builderProvider = new DocumentMKBuilderProvider();
+
+    private Clock virtualClock;
+    private DocumentNodeStore ns1;
+    private DocumentNodeStore ns2;
+
+    private ControllableMemoryDocumentStore store;
+    
+    @Before
+    public void setup() throws Exception {
+        int createOrUpdateBatchSize = 1;
+        setup(createOrUpdateBatchSize);
```

Comment:
we can inline this parameter and remove local variable.

Reply:
yes, looks a bit ugly indded, I've fixed this now in https://github.com/apache/jackrabbit-oak/pull/950/commits/39132a3385dd1d30d0e89d091d5f75e0b27f4523